### PR TITLE
Device Bootstrapper

### LINKED ISSUE #24667: Support creating device with teleport start --bootstrap
### What would you like Teleport to do?
Support creating device with `teleport start --bootstrap <dump from tctl get all>.yml`. More details on docs: https://goteleport.com/docs/management/operations/backup-restore/#example-of-backing-up-and-restoring-a-cluster-1

### What problem does this solve?
Restore devices with backup `.yml` file created with `tctl get all` command.

### If a workaround exists, please include it.
The current workaround is to use `tctl get devices > devices.yaml`, `tctl create devices.yml`.

### PR DESCRIPTION
**Pull request by flyinghermit**
Adds device bootstrapper which allows for registering device resources bootstrap logic in `e`, which can be used in `oss`. The logic is similar to existing [resource marshalers](https://github.com/gravitational/teleport.e/blob/master/tool/modules/resources.go) in `e`. 

### How it works:
1. Registers bootstrap func defined in `e`: corresponding PR https://github.com/gravitational/teleport.e/pull/1210
2. Runs after `local.CreateResources()`

Fixes: https://github.com/gravitational/teleport/issues/24667
Depends on https://github.com/gravitational/teleport/pull/24963 and https://github.com/gravitational/teleport.e/pull/1563.

### CODE REVIEW

File: lib/services/resource.go [hunk 1]
```diff
@@ -638,3 +640,41 @@ func (u *UnknownResource) UnmarshalJSON(raw []byte) error {
...
+			filteredResources.LocalResources = append(filteredResources.LocalResources, r)
+		}
+	}
+	return
+}
+
+type BootstrapDeviceResource func(context.Context, backend.Backend, *types.DeviceV1) error
+
+var eDeviceBootstrapper BootstrapDeviceResource
+var eDeviceBootsrapMutex sync.RWMutex
```
Comment (AntonAM) [inline lib/services/resource.go:664]: Nit: mutext on top of what it protects
```suggestion
var eDeviceBootsrapMutex sync.RWMutex
var eDeviceBootstrapper BootstrapDeviceResource
```

File: lib/services/resource.go [hunk 2]
```diff
@@ -638,3 +640,41 @@ func (u *UnknownResource) UnmarshalJSON(raw []byte) error {
...
+
+func RegisterDeviceBootstrapper(kind string, bootstrapper BootstrapDeviceResource) {
+	eDeviceBootsrapMutex.Lock()
+	defer eDeviceBootsrapMutex.Unlock()
+	eDeviceBootstrapper = bootstrapper
+}
+
+func BootstrapDevices(ctx context.Context, b backend.Backend, devices []types.DeviceV1) error {
+	for _, dev := range devices {
+		if err := eDeviceBootstrapper(ctx, b, &dev); err != nil {
```
Comment (AntonAM) [inline lib/services/resource.go:674]: Should we use mutex to protect this usage as well?

File: lib/services/resource.go [hunk 3]
```diff
@@ -638,3 +640,41 @@ func (u *UnknownResource) UnmarshalJSON(raw []byte) error {
...
+		case *types.DeviceV1:
+			filteredResources.DeviceResources = append(filteredResources.DeviceResources, *r)
+		default:
+			filteredResources.LocalResources = append(filteredResources.LocalResources, r)
+		}
+	}
+	return
+}
+
+type BootstrapDeviceResource func(context.Context, backend.Backend, *types.DeviceV1) error
```
Comment (codingllama) [inline lib/services/resource.go:661]: Pass all devices at once.

```suggestion
type BootstrapDeviceResource func(context.Context, backend.Backend, []*types.DeviceV1) error
```

File: lib/services/resource.go (hunk 1 above)
Comment (codingllama) [reply lib/services/resource.go:664]: I would drop the mutex and document, on RegisterDeviceBootstrapper, that it must be called before a Teleport server is initialized.

File: lib/services/resource.go (hunk 2 above)
Comment (codingllama) [inline lib/services/resource.go:674]: Handle `eDeviceBootstrapper == nil`.

File: lib/services/resource.go [hunk 4]
```diff
@@ -638,3 +640,41 @@ func (u *UnknownResource) UnmarshalJSON(raw []byte) error {
...
+
+var eDeviceBootstrapper BootstrapDeviceResource
```
Comment (codingllama) [inline lib/services/resource.go:663]: Drop the "e" prefix, seems like a needless assumption here.

```suggestion
var deviceBootstrapper BootstrapDeviceResource
```

File: lib/services/resource.go [hunk 5]
```diff
@@ -638,3 +640,41 @@ func (u *UnknownResource) UnmarshalJSON(raw []byte) error {
 	copy(u.Raw, raw)
 	return nil
 }
+
+type bootstrapResourcesType struct {
+	LocalResources  []types.Resource
+	DeviceResources []types.DeviceV1
+}
+
+func FilterBootstrapResourcesByType(resources ...types.Resource) (filteredResources bootstrapResourcesType) {
```
Comment (codingllama) [inline lib/services/resource.go:649]: Move this to lib/auth and make it private, there's little reason to have it here.

File: lib/services/resource.go (hunk 3 above)
Comment (codingllama) [inline lib/services/resource.go:661]: Please add a godoc.

File: lib/services/resource.go [hunk 6]
```diff
@@ -638,3 +640,41 @@ func (u *UnknownResource) UnmarshalJSON(raw []byte) error {
...
+var eDeviceBootstrapper BootstrapDeviceResource
+var eDeviceBootsrapMutex sync.RWMutex
+
+func RegisterDeviceBootstrapper(kind string, bootstrapper BootstrapDeviceResource) {
+	eDeviceBootsrapMutex.Lock()
+	defer eDeviceBootsrapMutex.Unlock()
+	eDeviceBootstrapper = bootstrapper
+}
+
+func BootstrapDevices(ctx context.Context, b backend.Backend, devices []types.DeviceV1) error {
```
Comment (codingllama) [inline lib/services/resource.go:672]: Please add a godoc.

**Review by codingllama — CHANGES_REQUESTED** [review]

**Comment by sshahcodes** [comment]
Closing this as the PR has been idle for so long, lags far behind the master branch, and I don't have any immediate plans for this right now.

Future note added in `e` counterpart: https://github.com/gravitational/teleport.e/pull/1210#issuecomment-1706554477

**Pull request closed without merging**